crossplane-infra-access/README.md

```diff
@@ -0,0 +1,50 @@
+# Crossplane infrastructure access
+
+## Intro
+
+Team Honeybadger is deploying Crossplane to MCs. Crossplane can be used to manage infrastructure on the cloud provider side by both us and out customers.
+
+In order to do that, Crossplane has to be able to authenticate and authorize against the infrastructure provider. The access model that Crossplane uses depends on the user. By default, Crossplane doesn't require nor provide any access info. The administrator of Crossplane can create as many access secrets as he/she wants and can grant access to them to other Kubernetes users using the standard RBAC configuration.
+
+Example:
+1. Crossplane admin logins to an AWS account. In IAM, it creates two roles:
+  - `s3-access` that grants permissions to use S3 only,
+  - `db-access` that allows usage of RDS,
+2. For both roles, the admin creates access keys. Then, access keys are saved as secrets in the kubernetes API.
```

Comment:
I think that we want to move away from any manually generated access keys as much as possible, as the general approach in AWS itself (the guidelines) and also our customer requirements are often that we do not use accesskeys but rather have roles and assume the role.

For that case we deeply IRSA which can leverage assuming specific roles based on the pod needs

Reply:
+1 no hardcoded secrets ever

Reply:
+ 1 for not hardcoding secrets / using access keys.

For reference, the community AWS provider's provider config (the thing used to let the provider controller access to the cloud provider for a given resource (as this can be different for every resource / composition / whatever)): https://doc.crds.dev/github.com/crossplane/provider-aws/aws.crossplane.io/ProviderConfig/v1beta1@v0.31.0

It can assume a given role / web identity, but a secret is needed that has some sort of credentials / tokens.

Something maybe that could regenerate this token periodically would be nice. As of now I don't know if IRSA could work here or now as I don't know it very well.

Reply:
IRSA could work, according to: https://github.com/crossplane-contrib/provider-aws/blob/bf4088121aa87877cb969c5de3e1b287d6f853ca/AUTHENTICATION.md#using-iam-roles-for-serviceaccounts

Reply:
but IIRC for assuming a role you have to be an AWS user (so credentials for origin account). Like our aws operator, it has user credentials for our base AWS account and it assumes the  roles in our customer accounts, right?

Reply:
No,  we don't need, IRSA does it without the need for credentials, similar to what KIAM does

Reply:
https://github.com/crossplane-contrib/provider-aws/blob/bf4088121aa87877cb969c5de3e1b287d6f853ca/AUTHENTICATION.md#using-iam-roles-for-serviceaccounts

looks like it should work

Reply:
> However the problem must be solved for other cloud providers as well. We support AWS, GCP and Azure right now with Crossplane.

All providers have similar functionality based on the cloud roles so this should be similar solution across all providers

Reply:
and FYI we have a customer who doesn't allow the use and manual creation of any static Access/Secret keys and we are only allowed to use role assuming so anyway  any approach not leveraging IRSA will not work there and will be denied

Reply:
The problem is that IRSA is not a good solution as well: it can use only a single kubernetes Role. More precisely, we can bind the SA that `aws-provider` runs with with an IAM role, but it has to be a single SA and thus a single IAM role. We can't make fine-grained separate `ProviderConfigs`. This means that this IAM role we bind with IRSA needs to be almost "aws admin", or at least have all the required permissions for a given MC.

Reply:
AFAIK IRSA (as the name suggests is access automation for Service Accounts) and those SAs can then be used separately (in form of secrets for example), they do not need to be part of the deployment, even if that's the usual process.